contracts/eosio.system/src/wps.cpp

```diff
@@ -0,0 +1,774 @@
+
+        auto itr = _committees.find(committeeman.value);
+
+        check(itr != _committees.end(), "Account not found in committee table");
+        _committees.erase(itr);
+    }
+
+    void system_contract::rejectfund(name committeeman, name proposer, const string& reason){
+        require_auth(committeeman);
+
+        check(is_account(proposer), "Proposal creator does not exist");
+
+        check(reason.size() > 0, "must provide a brief reason");
+        check(reason.size() < 256, "reason is too long");
+
+        auto itr = _committees.find(committeeman.value);
+        // verify that the committee is on committee table
+        check(itr != _committees.end(), "Account not found in committee table");
+
+        auto itr_proposal = _proposals.find(proposer.value);
+        check(itr_proposal != _proposals.end(), "Proposal not found in proposal table");
+
+        check((*itr_proposal).committee == (*itr).committeeman || (*itr).is_oversight, "Committee is not associated with this proposal");
+        check((*itr_proposal).status == PROPOSAL_STATUS::APPROVED, "Proposal::status is not PROPOSAL_STATUS::APPROVED");
+
+        _proposals.modify(itr_proposal, same_payer, [&](auto& _proposal){
+            _proposal.status = PROPOSAL_STATUS::REJECTED;
+        });
+    }
+
+    double stake2vote2( int64_t staked ) {
+        // From voting.cpp
+        /// TODO subtract 2080 brings the large numbers closer to this decade
+        double weight = int64_t( (current_time_point().sec_since_epoch() - (block_timestamp::block_timestamp_epoch / 1000)) / (seconds_per_day * 7) )  / double( 13 );
+        return double(staked) * std::pow( 2, weight );
+    }
+
+    void system_contract::voteproposal( const name& voter_name, const std::vector<name>& proposals ) {
+        require_auth( voter_name );
+        // vote_stake_updater( voter_name );
+        update_wps_votes( voter_name, proposals, true );
+    }
+
+    void system_contract::update_wps_votes( const name& voter_name, const std::vector<name>& proposals, bool voting){
+        //validate input
+        check( proposals.size() <= 1, "attempt to vote for too many proposals" );
+        for( size_t i = 1; i < proposals.size(); ++i ) {
+            check( proposals[i-1] != proposals[i], "proposal votes must be unique" );
+        }
+
+        auto voter = _voters.find( voter_name.value );
+        check( voter != _voters.end(), "user must stake before they can vote" ); /// staking creates voter object
+
+        auto wpsvoter = _wpsvoters.find( voter_name.value );
+
+        /**
+         * The first time someone votes we calculate and set last_vote_weight, since they cannot unstake until
+         * after total_activated_stake hits threshold, we can use last_vote_weight to determine that this is
+         * their first vote and should consider their stake activated.
+         */
+        if( voter->last_vote_weight <= 0.0 ) {
```

Comment:
I don't think we should be messing with the `_gstate.total_activated_stake` value in 2 different places. This function should have as little interaction with the regular voter logic and associated global state as possible. Better to assert that the activated stake has gone beyond the threshold. Ie replace this block with:
```
check( _gstate.total_activated_stake >= min_activated_stake,
             "cannot update wps votes until the chain is activated (at least 15% of all tokens participate in voting)" );
```

Reply:
Agreed, thanks for this